minitask2/akhmad/behaviour.py

```diff
@@ -0,0 +1,196 @@
+left_turn_cmd.linear.x = 0
+left_turn_cmd.angular.z = radians(30)
+
+right_turn_cmd = Twist()
+right_turn_cmd.linear.x = 0
+right_turn_cmd.angular.z = radians(-30)
+
+save_position = True
+last_pose_theta = None
+last_pose_x = None
+last_pose_y = None
+
+continue_right = False
+continue_left = False
+
+front = None
+right_front = None
+right_back = None
+right = None
+back = None
+left_back = None
+left_front = None
+left = None
+
+is_wall_found = False
+
+
+class Behaviour():
+    def __init__(self):
+        self.behaviour = input("Enter behaviour mode '0' for random walk '1' for right hand wall following: ")
+        rospy.init_node('behaviour')
+
+        # Subscribers
+        self.laser_sub = rospy.Subscriber('/scan', LaserScan, self.scan_callback)
+        self.odom_sub = rospy.Subscriber('/odom', Odometry, self.odom_callback)
+
+        # Publisher
+        self.cmd_vel_pub = rospy.Publisher('/cmd_vel', Twist, queue_size=3)
+
+        self.rate = rospy.Rate(10)
+        self.current_pose_theta = None
+        self.current_pose_x = None
+        self.current_pose_y = None
+        rospy.spin()
+
+    def publish_command(self, msg):
+        self.cmd_vel_pub.publish(msg)
+
+    def turn_left(self):
+        global continue_left, continue_right
+        continue_right = False
+        continue_left = True
+        self.publish_command(left_turn_cmd)
+        print("Avoiding obstacle by turning left")
+
+    def turn_right(self):
+        global continue_left, continue_right
+        continue_right = True
+        continue_left = False
+        self.publish_command(right_turn_cmd)
+        print("Avoiding obstacle by turning right")
+
+    def right_hand_wall_follow(self):
+        global is_wall_found
+
+        if not is_wall_found:
+            if all(i >= 0.5 for i in front):
+                self.publish_command(move_cmd)
+            else:
+                is_wall_found = True
+                self.publish_command(Twist())
+        else:
+            if min(right) <= 0.5 and min(front) <= 0.5:
+                self.publish_command(left_turn_cmd)
+                print("Obstacles on front and right - turning left")
+            elif min(right) <= 0.5 and min(front) >= 0.5:
+                self.publish_command(move_cmd)
+                print("No obstacles in front but on right - move forward")
+            elif min(right) >= 0.5 and min(front) <= 0.5:
+                self.publish_command(left_turn_cmd)
+                print("No obstacles on right but in front - turning left")
+            elif min(right) >= 0.5 and min(front) >= 0.5:
+                if min(right_front) <= 0.5:
+                    self.publish_command(right_turn_cmd)
+                else:
+                    self.publish_command(move_cmd)
+                print("No obstacles on right and in front - sliding right")
+
+    def obstacle_avoidance(self):
+        # continue_right and continue_left is used to
+        # keep turning in same direction as robot started
```

Comment:
*Potentially a worse solution*, but I just always turned in one direction to simplify this behaviour.

Reply:
What I did here is not necessary nor it is asked in ministask, but I just wanted to minimize the turning time.

Reply:
Okay, If it works well then I think its a good improvement!